content/blog/2016/2016-08-29-sauce-pipeline.adoc

```diff
@@ -0,0 +1,271 @@
+
+I ran this job a few times to get the JUnit report to show a trend graph.
+
+image::/images/post-images/2016-08-29/pipeline-report.png[Pipeline Report for "sauce-pipeline" Branch, role="center"]
+
+This sample app generates the `SauceOnDemandSessionID` for each test, enabling the Jenkins Sauce OnDemand Plugin's result publisher to link results to details Sauce Labs captured during the run.
+
+image::/images/post-images/2016-08-29/pipeline-results.png[Sauce Test Results, role="center"]
+
+image::/images/post-images/2016-08-29/sauce-details.png[Sauce Test Result Details, role="center"]
+
+
+=== Adding Platforms
+Next I added some more platforms to my matrix.
+This is a perfect example of the power of pipeline as code.
+Adding new platforms required changing both the pipeline and the Nightwatch.js configuration file.
+Since my pipeline was now checked in, I could change both as part of one commit and be assured my pipeline and test framework wouldn't get out of sync.
+
+I added three new environments to `nightwatch.json`:
+
+[source,js]
+----
+"test_settings" : {
+  "default": { /*----8<----8<----8<----*/ },
+  "chrome": { /*----8<----8<----8<----*/ },
+
+  "firefox": {
+    "desiredCapabilities": {
+      "platform": "linux",
+      "browserName": "firefox",
+      "version": "latest"
+    }
+  },
+  "ie": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "internet explorer",
+      "version": "latest"
+    }
+  },
+  "edge": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "MicrosoftEdge",
+      "version": "latest"
+    }
+  }
+}
+----
+
+And I modified my `Jenkinsfile` to call them:
+
+[source,groovy]
+----
+//----8<----8<----8<----8<----8<----8<----
+sauceconnect(options: '', useGeneratedTunnelIdentifier: false, verboseLogging: false) {
+    def configs = [ // <1>
+        'chrome',
+        'firefox',
+        'ie',
+        'edge'
+    ].join(',')
+    // Run selenium tests using Nightwatch.js
+    sh "./node_modules/.bin/nightwatch -e ${configs} --test tests/guineaPig.js" // <2>
+} //----8<----8<----8<----8<----8<----8<----
+----
+<1> Using an array to improve readability and make it easy to add more platforms later.
+<2> Changed from single-quoted string to double-quoted to support variable substitution.
+
+[WARNING]
+====
+Test frameworks have bugs, too. Nightwatch.js (v0.9.8) generates incomplete JUnit files,
+reporting results without enough information in them to distinguish between platforms.
+I implemented a fix for it and
+link:https://github.com/nightwatchjs/nightwatch/pull/1160[submitted a PR] to Nightwatch.js.
+This blog shows output with that fix applied locally.
+
+====
+
+As expected, Jenkins picked up the new pipeline and ran Nightwatch.js on four platforms.
+Nightwatch.js was already configured to use multiple worker threads to run tests against those platforms in parallel, and
+my Sauce Labs account supported running them all at the same time.
```

Comment:
There's an underlying point here which I think you could make more explicit: running these tests against multiple platforms in Sauce Labs in parallel means you can run a lot of acceptance tests faster.